### ISSUE TITLE
Question: Creating first widget, can't get to show window, throws exception.

### ISSUE THREAD

**Issue by klong15** [microsoft/XboxGameBarSamples#54]
I've been following along with the development guide, and I've got it to activate properly, but the window doesn't show. When debugging, I get this error when I try to open the widget:

`Exception thrown at 0x77CB4402 (KernelBase.dll) in DailyTracker.exe: WinRT originate error - 0x80040155 : 'Failed to find proxy registration for IID: {FD037AC2-5D3A-438D-BD53-88F0B6B0F399}.'.`

Not quite sure what do try at this point. I'm pretty new to this environment too. My project is in C++/CX if that matters. The guide has been in C++/CLI I think? So I've had to modify the code a bit to get different parts to just compile. 

The error makes me think the proxy/stub registration I had to add to the manifest is wrong somehow (no idea what this actually does), but I've done exactly what the guide said to add:
````
<Extensions>
    <Extension Category="windows.activatableClass.proxyStub">
      <ProxyStub ClassId="00000355-0000-0000-C000-000000000046">
        <Path>Microsoft.Gaming.XboxGameBar.winmd</Path>


        <!-- include when using SDK version 5.1+-->
        <Interface Name="Microsoft.Gaming.XboxGameBar.Private.IXboxGameBarWidgetAuthHost" InterfaceId="DC263529-B12F-469E-BB35-B94069F5B15A" />
        <Interface Name="Microsoft.Gaming.XboxGameBar.Private.IXboxGameBarWidgetControlHost" InterfaceId="C309CAC7-8435-4082-8F37-784523747047" />
        <Interface Name="Microsoft.Gaming.XboxGameBar.Private.IXboxGameBarNavigationKeyCombo" InterfaceId="5EEA3DBF-09BB-42A5-B491-CF561E33C172" />
        <Interface Name="Microsoft.Gaming.XboxGameBar.Private.IXboxGameBarWidgetActivatedEventArgsPrivate" InterfaceId="782535A7-9407-4572-BFCB-316B4086F102" />
        <Interface Name="Microsoft.Gaming.XboxGameBar.Private.IXboxGameBarWidgetHost" InterfaceId="5D12BC93-212B-4B9F-9091-76B73BF56525" />
        <Interface Name="Microsoft.Gaming.XboxGameBar.Private.IXboxGameBarWidgetPrivate" InterfaceId="22ABA97F-FB0F-4439-9BDD-2C67B2D5AA8F" />

        <!-- include when using SDK version 5.3+-->
        <Interface Name="Microsoft.Gaming.XboxGameBar.Private.IXboxGameBarWidgetHost2" InterfaceId="28717C8B-D8E8-47A8-AF47-A1D5263BAE9B" />
        <Interface Name="Microsoft.Gaming.XboxGameBar.Private.IXboxGameBarWidgetPrivate2" InterfaceId="B2F7DB8C-7540-48DA-9B46-4E60CE0D9DEB" />

      </ProxyStub>
    </Extension>
  </Extensions>

````
I've got version 5.1.200402003 of XboxGameBar installed, so shouldn't need the last two InterfaceId lines. But I've tried it with them just to make sure, and still get the error.

Sorry if this isn't the right spot for this, but its the only discussion area I could find!

**Comment by hamiltonia** [comment]
The last two lines are for when you're using the latest 5.3 pre-release SDK. Your Game Bar version is 5.1 so it's the release version. What version of the NuGet package / SDK are you using? With your version of Game Bar you should be using the latest release (5.1.200402003). 

Did you step through your OnActivate and verify everything is being setup correctly there?

**Comment by klong15** [comment]
Yep, OnActivated() gets hit, and I stepped through it. It goes inside of the IsLaunchActivation if check, and created the frame and activated it. Here is what my OnActivated() looks like.

```
void App::OnActivated(IActivatedEventArgs ^ e) 
{
    XboxGameBarWidgetActivatedEventArgs ^ widgetArgs = nullptr;
    if (e->Kind == ActivationKind::Protocol)
    {
        IProtocolActivatedEventArgs^ protocolArgs = dynamic_cast<IProtocolActivatedEventArgs^>(e);
        
        if (protocolArgs)
        {
            String^ scheme = protocolArgs->Uri->SchemeName;
            if (scheme->Equals("ms-gamebarwidget")) {
                widgetArgs = dynamic_cast<XboxGameBarWidgetActivatedEventArgs^>(e);
            }
        }
    }
    if (widgetArgs)
    {
        if (widgetArgs->IsLaunchActivation)
        {
            // Game Bar Widget activation work here
            auto rootFrame = ref new Frame();
            rootFrame->NavigationFailed += ref new Windows::UI::Xaml::Navigation::NavigationFailedEventHandler(this, &App::OnNavigationFailed);
            
            Window::Current->Content = rootFrame;
            rootFrame->Navigate(TypeName(MyPage::typeid));
            Window::Current->Activate();
        }
        else
        {
            // Repeat activation for this widget. Ignore, or respond to URI payload
        }
    }
}
```
The code steps through to the Activate() call. But while stepping through, the exception gets thrown when I do the dynamic_cast for widgetArgs.

**Comment by vidager** [comment]
Hi klong15,

This message in the debug output is a red-herring:

Exception thrown at 0x77CB4402 (KernelBase.dll) in DailyTracker.exe: WinRT originate error - 0x80040155 : 'Failed to find proxy registration for IID: {FD037AC2-5D3A-438D-BD53-88F0B6B0F399}.'.

This means COM couldn't find a proper proxy/stub dll to marshal the interface of that GUID. However, after KernelBase.dll spits out that error, COM proceeds to search for marshaling info via a little known COM feature called Metadata Based Marshaling (MBM).  That's where COM looks for marshaling info in the WinMD file (we payload the winmds with the nuget).  So the issue with your widget not loading lies elsewhere.

Can you clarify what other (if any) exceptions or error you are hitting?  If so, do they reproduce with the CX sample we provide in the sample repo here?

**Comment by vidager** [comment]
Side note: Our guide is written for C++/WinRT not C++/CLI.  C++/WinRT is the replacement for C++/CX for consuming WinRT APIs.  C++/WinRT is a fully C++ compliant header library with allows apps to consume WinRT.  
https://docs.microsoft.com/en-us/windows/uwp/cpp-and-winrt-apis/

With that said, you can freely write widgets in .NET, C++/CX, or C++/WinRT.  We'll typically only demonstrate advanced SDK features in C# or C++/WinRT, but we are more than happy to help with CX as needed.  Let us know how your next steps go on your widget and we're happy to help!

**Comment by klong15** [comment]
Ah ok. Well I just switched to using C++/WinRT then. Kind of new to the whole environment, and CX was the only type of project to create in the project creation wizard.

I still had to tweak some of the code from the guide to get it to work, but I got a widget to show! The only little issue I couldn't get passed was when trying to include a settings page with the <SettingsSupported> tag. The guide says to include another <uap3:AppExtension> block for the settings, but I get an "App manifest validation error" on the second <uap3:AppExtension> block. Here is my appxmanifest. It throws the validation error about half way down (marked with a comment).

```
<Applications>
    <Application Id="App" Executable="$targetnametoken$.exe" EntryPoint="DailyTracker.App">
      <uap:VisualElements DisplayName="DailyTracker" Description="Project for a single page C++/WinRT Universal Windows Platform (UWP) app with no predefined layout"
        Square150x150Logo="Assets\Square150x150Logo.png" Square44x44Logo="Assets\Square44x44Logo.png" BackgroundColor="transparent" AppListEntry="none">
        <uap:DefaultTile Wide310x150Logo="Assets\Wide310x150Logo.png">
        </uap:DefaultTile>
        <uap:SplashScreen Image="Assets\SplashScreen.png" />
      </uap:VisualElements>
      <Extensions>
        <uap3:Extension Category="windows.appExtension">
          <uap3:AppExtension Name="microsoft.gameBarUIExtension"
                             Id="klong0815"
                             DisplayName="Daily Tracker"
                             Description="Temp description"
                             PublicFolder="GameBar\Ext1">
            <uap3:Properties>
              <GameBarWidget Type="Standard">
                <HomeMenuVisible>true</HomeMenuVisible>
                <PinningSupported>true</PinningSupported>
                <SettingsSupported AppExtensionId="klong0815settings" />
                <Window>
                  <Size>
                    <Height>300</Height>
                    <Width>400</Width>
                    <MinHeight>150</MinHeight>
                    <MinWidth>200</MinWidth>
                    <MaxHeight>350</MaxHeight>
                    <MaxWidth>450</MaxWidth>
                  </Size>
                  <ResizeSupported>
                    <Horizontal>true</Horizontal>
                    <Vertical>true</Vertical>
                  </ResizeSupported>
                </Window>
              </GameBarWidget>
            </uap3:Properties>
          </uap3:AppExtension>
          <!--IT ERRORS ON THE LINE BELOW-->
          <uap3:AppExtension Name="microsoft.gameBarUIExtension"
                             Id="klong0815settings"
                             DisplayName="Daily Tracker Settings"
                             Description="My settings"
                             PublicFolder="GameBar\Ext1">
            <uap3:Properties>
              <GameBarWidget Type="Settings">
                <Window>
                  <Size>
                    <Height>300</Height>
                    <Width>300</Width>
                  </Size>
                  <ResizeSupported>
                    <Horizontal>false</Horizontal>
                    <Vertical>false</Vertical>
                  </ResizeSupported>
                </Window>
              </GameBarWidget>
            </uap3:Properties>
          </uap3:AppExtension>
        </uap3:Extension> 
      </Extensions>
    </Application>
  </Applications>
```
Thank you for the help getting me started though! I've got plenty to get started and tinker with stuff. If my issue with the settings option is blatantly obvious issue, feel free to let me know,

**Comment by hamiltonia** [comment]
Did you add "uap3" to the top of your manifest? It's needed as both a namespace and in the `IgnorableNamespaces` section.

```
<Package xmlns="..."
         xmlns:mp="..."
         xmlns:uap="..."
         xmlns:uap3="http://schemas.microsoft.com/appx/manifest/uap/windows10/3"
         IgnorableNamespaces="uap mp uap3">
```

**Comment by vidager** [comment]
Hi Klong15,

Looking at your manifest entries I don't see anything obvious right now.  You might do a diff against our Settings Sample on this github repo to double check for any minor syntax errors.

Also, if you share the specific error it might help.

Thanks,
Ben

**Comment by klong15** [comment]
Turns out the <uap3:AppExtension> for the Settings needed to be wraped in its own <uap3:Extension> block. The guide has it as a child of the same <uap3:Extension> of the main widget. But in the samples has it the other way.

So now I've set up my project to be similar to the WidgetSettingsSample project.

But another hurdle. Clicking the settings button doesn't open up my settings widget.  It never hits App::OnActivated() when I click the settings button. I've even copied the appxmanifest file completely from the sample project, and it still doesn't hit App:OnActivated(). Is there another area other than the appxmanifest to look at to get the settings button to work properly?
